### PR TITLE
Status updates sent to the Aria live region as the user completes eac…

### PR DESCRIPTION
**Pull request by staubina**
…h section of an ORA2 problem

### Description

This PR focuses on Sub Task https://openedx.atlassian.net/browse/TNL-6241

### Sandbox
- [x] Build a sandbox for your branch and add a link here
https://staubina-ora.sandbox.edx.org/

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens 
- [x] Code review: @efischer19 
- [x] Accessibility review: @cptvitamin 

### Post-review
- [x] Rebase and squash commits

FYI: @jlajoie

### CODE REVIEW

File: openassessment/templates/openassessmentblock/peer/oa_peer_waiting.html [hunk 1]
```diff
@@ -13,7 +13,7 @@
 {% block title %}
 <span class="step__status">
   <span class="step__status__value">
-    <span class="copy">{{review_num}}
+    <span class="copy">
```
Comment (cahrens) [inline openassessment/templates/openassessmentblock/peer/oa_peer_waiting.html]: Was this just a bug?

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 2]
```diff
@@ -68,6 +69,67 @@ OpenAssessment.BaseView.prototype = {
     },
 
     /**
+     * Checks the rendering status of the views.
+     * @returns {boolean} true if any step's view is still loading.
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Learner-facing views? I note that messageView, leaderboardView, and staffAreaView are not listed here.

If new views are added, it would be helpful to understand the contract of this method, so as to know whether or not the new views should be added here as well.

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 3]
```diff
@@ -68,6 +69,67 @@ OpenAssessment.BaseView.prototype = {
...
+            this.srStatusUpdates.push(text);
+        }
+        if (!this.isLoading() && this.srStatusUpdates.length > 0){
+            window.SR.readTexts(this.srStatusUpdates);
+            this.srStatusUpdates = []
+        }
+    },
+
+    /**
+     * Retrieves and returns the current status of a given step.
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: It is not clear to me how this method differs from getAssessmentGradeStatus.

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 4]
```diff
@@ -68,6 +69,67 @@ OpenAssessment.BaseView.prototype = {
...
+     *
+     * @param stepID - The id of the Step to retrieve grade status for.
+     * @param usageID - The Usage ID of the current grade status.
+     * @returns {string} - the current grade status.
+     */
+    getAssessmentGradeStatus: function(stepID, usageID) {
+        var block_id = "[data-usage-id='" + usageID + "'] ";
+        var css_base = block_id + stepID + " .step__header .step__title ";
+        var css_string_title = css_base + ".step__label";
+        var css_string_status = css_base + ".grade__value";
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: It looks like this line is the only difference between this method and getAssessmentStatus. Can refactor and share code.

File: openassessment/xblock/static/js/src/lms/oa_grade.js [hunk 5]
```diff
@@ -20,17 +22,31 @@ OpenAssessment.GradeView.prototype = {
...
+                if (typeof usageID !== 'undefined' &&
+                    $(stepID, view.element).hasClass("is--showing")){
+                    view.announceStatusChange = true;
+                }
+                if (view.announceStatusChange) {
+                    // Update the SR and announce the status change
+                    view.baseView.announceStatusChangeToSR(view.baseView.getAssessmentGradeStatus(stepID, usageID));
+                    view.announceStatusChange = false;
+                } else {
+                    view.baseView.announceStatusChangeToSR();
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: I'm not clear about the logic of why this method is called with no arguments.

File: openassessment/xblock/static/js/src/lms/oa_grade.js [hunk 6]
```diff
@@ -20,17 +22,31 @@ OpenAssessment.GradeView.prototype = {
...
                 if (typeof usageID !== 'undefined' &&
                     !$(stepID, view.element).hasClass("is--unfinished") &&
                     !$(stepID, view.element).hasClass("is--unstarted") &&
                     !$(stepID, view.element).hasClass("is--waiting--staff")) {
                     $("[id='oa_grade_" + usageID + "']", view.element).focus();
+                    view.announceStatusChange = true;
+                }
+                if (typeof usageID !== 'undefined' &&
+                    $(stepID, view.element).hasClass("is--showing")){
+                    view.announceStatusChange = true;
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: Can you explain why in this case we do want to announce a status change, but we don't want to focus (a comment in the code would be helpful)?

File: openassessment/xblock/static/js/src/lms/oa_peer.js [hunk 7]
```diff
@@ -26,14 +28,28 @@ OpenAssessment.PeerView.prototype = {
     load: function(usageID) {
         var view = this;
         var stepID = ".step--peer-assessment";
+
+        view.isRendering = true;
         this.server.render('peer_assessment').done(
             function(html) {
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_peer.js]: The duplication between the various JS classes is painful. Is there any way to create a shared "done" handler?

File: openassessment/xblock/static/js/src/lms/oa_peer.js [hunk 8]
```diff
@@ -123,9 +152,14 @@ OpenAssessment.PeerView.prototype = {
                 // Override default form submission
                 eventObject.preventDefault();
 
+
+                //Status will change in update announce it to the Screen Reader after Render
+                view.announceStatusChange = true;
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_peer.js]: I'm not clear why this explicit call is necessary (as opposed to the handling above in the render callbacks.

File: openassessment/xblock/static/js/src/lms/oa_peer.js [hunk 9]
```diff
@@ -14,6 +14,8 @@ OpenAssessment.PeerView = function(element, server, baseView) {
     this.server = server;
     this.baseView = baseView;
     this.rubric = null;
+    this.isRendering = false;
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_peer.js]: Should this be pushed down into the baseView?

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 10]
```diff
@@ -24,6 +24,8 @@ OpenAssessment.ResponseView = function(element, server, fileUploader, baseView,
     this.autoSaveTimerId = null;
     this.data = data;
     this.fileUploaded = false;
+    this.announceStatus = false;
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: Why announceStatus vs. announceStatusChange?

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 11]
```diff
@@ -423,9 +435,13 @@ OpenAssessment.ResponseView.prototype = {
     moveToNextStep: function() {
         var baseView = this.baseView;
         var usageID = baseView.getUsageID();
+        var view = this;
+
         this.load(usageID);
         baseView.loadAssessmentModules(usageID);
 
+        view.announceStatus = true;
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: Why is this set here vs. the render callback (like in the other views)?

**Comment by cahrens** [comment]
@staubina it's not clear to me yet how you are determining when status text needs to be read vs. not. And I feel like there should be some way to share code, but it seems that there are slight variations in some of the views that I don't understand.

Let's talk through this in person tomorrow morning.

File: openassessment/templates/openassessmentblock/peer/oa_peer_waiting.html (hunk 1 above)
Comment (staubina) [inline openassessment/templates/openassessmentblock/peer/oa_peer_waiting.html]: I believe so, I need to verify with Ux.

File: openassessment/xblock/static/js/src/lms/oa_base.js (hunk 4 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Yes. In the next PR.

File: openassessment/xblock/static/js/src/lms/oa_grade.js (hunk 6 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: I was avoiding changing existing functionality, there is a specific case on line 37 that has the focus.  I will test this case though to see if it would be valuable to focus here as well.

File: openassessment/xblock/static/js/src/lms/oa_grade.js (hunk 5 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: The views can complete their render in a different order, this method without parameters checks to see if the views are still rendering and if they are done adds the text to the aria live region.

File: openassessment/xblock/static/js/src/lms/oa_peer.js (hunk 8 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_peer.js]: This call sets the flag on a submit to announce the status change after the render is complete.

File: openassessment/xblock/static/js/src/lms/oa_response.js (hunk 10 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: Typo.

File: openassessment/xblock/static/js/src/lms/oa_response.js (hunk 11 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: This is on submit.  Then when the page refreshes and renders from the submit it will announce the status.

File: openassessment/xblock/static/js/src/lms/oa_base.js (hunk 2 above)
Comment (staubina) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: messageView, I am still looking at.

leaderboardView we discussed.  I do not believe it needs an aria-live update.

I will take a look at staffAreaView, My initial thought was that it would not need an aria live region update.  That might be wrong though.

File: openassessment/xblock/static/js/src/lms/oa_grade.js [hunk 12]
```diff
@@ -20,17 +22,31 @@ OpenAssessment.GradeView.prototype = {
...
         var stepID = ".step--grade";
+        view.isRendering = true;
         this.server.render('grade').done(
             function(html) {
                 // Load the HTML and install event handlers
                 $(stepID, view.element).replaceWith(html);
                 view.server.renderLatex($(stepID, view.element));
+                view.isRendering = false;
                 view.installHandlers();
                 if (typeof usageID !== 'undefined' &&
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: Simplify to just is--showing if possible (for both focus change and announce status).

File: openassessment/xblock/static/js/src/lms/oa_self.js [hunk 13]
```diff
@@ -134,6 +148,7 @@ OpenAssessment.SelfView.prototype = {
         ).done(
             function() {
                 baseView.unsavedWarningEnabled(false, view.UNSAVED_WARNING_KEY);
+                view.announceStatusChange = true;
                 baseView.loadAssessmentModules(usageID);
                 view.load(usageID);
                 baseView.scrollToTop(".step--self-assessment");
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_self.js]: Remove this scroll?

File: openassessment/xblock/static/js/src/lms/oa_self.js [hunk 14]
```diff
@@ -134,6 +148,7 @@ OpenAssessment.SelfView.prototype = {
         ).done(
             function() {
                 baseView.unsavedWarningEnabled(false, view.UNSAVED_WARNING_KEY);
+                view.announceStatusChange = true;
                 baseView.loadAssessmentModules(usageID);
                 view.load(usageID);
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_self.js]: Can this load be removed?

File: openassessment/xblock/static/js/src/lms/oa_self.js (hunk 13 above)
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_self.js]: And announce errors in toggleActionError.

**Comment by staubina** [comment]
Notes from conversation with @cahrens 

Convert 

```
                if (view.announceStatus){
                    // Update the Aria-Live region with the current status.
                            
                   view.baseView.announceStatusChangeToSR(view.baseView.getAssessmentStatus(stepID, usageID));
                    view.announceStatus = false;
                } else {
                    view.baseView.announceStatusChangeToSR();
                }
```

To a single method in the base view and remove the `view.announceStatus = true;` line and use the conditional to know when to announce status on a load.

**Comment by staubina** [comment]
@cahrens I have made the changes that we discussed. I am going to convert this to a full PR once I get your feedback on the recent changes.

**Comment by staubina** [comment]
@cahrens Please wait to review, I just remembered the error messaging changes.  Also want to resolve some test failures I missed.

File: karma.conf.js [hunk 15]
```diff
@@ -26,6 +26,7 @@ module.exports = function(config) {
     // list of files / patterns to load in the browser
     files: [
       'lib/jquery.min.js',
+      'src/common/lib/xmodule/xmodule/js/common_static/js/src/accessibility_tools.js',
```
Comment (cahrens) [inline karma.conf.js]: This won't work-- this file lives in edx-platform. The lms and studio directories are subdirectories within ora itself.

The code either needs to be robust to that global function not existing, or we need to create our own sr div.

**Comment by staubina** [comment]
@cahrens This is now ready for you to take another look.  The Sandbox is also updated.

File: openassessment/xblock/static/js/src/lms/oa_grade.js [hunk 16]
```diff
@@ -20,18 +22,24 @@ OpenAssessment.GradeView.prototype = {
...
                 view.installHandlers();
-                if (typeof usageID !== 'undefined' &&
-                    !$(stepID, view.element).hasClass("is--unfinished") &&
-                    !$(stepID, view.element).hasClass("is--unstarted") &&
-                    !$(stepID, view.element).hasClass("is--waiting--staff")) {
-                    $("[id='oa_grade_" + usageID + "']", view.element).focus();
-                }
+
+                view.baseView.announceStatusChangeToSRandFocus(view.announceStatus,
+                    stepID,
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: Nit: I hate taking up all this room, and would put more than one arg on a line to minimize the number of lines.

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 17]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
     },
 
     /**
+     * Check if the Aria live region already exists, if it doesn't then add it.
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Nit: I'd reword this (the important part of the method is the reading of the text messages)--

Add the text messages to the Aria live region (first creating it if necessary).

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 18]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+        if ($readerFeedbackSelector.length === 0) {
+            $(this.element).append(
+                '<div class="sr reader-feedback" aria-live="polite"></div>'
+            );
+            $readerFeedbackSelector = $(readerFeedbackID);
+        }
+        $.each(texts, function(ids, value) {
+            htmlFeedback = htmlFeedback + '<p>' + value + '</p>\n';
+        });
+        $readerFeedbackSelector.html(htmlFeedback);
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Do you have any unit test coverage that these messages are added to the sr div?

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 19]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+
+    /**
+     * Checks the rendering status of the views that may require Screen Reader Status updates.
+     *
+     * The only views that should be added here are those that require Screen Reader updates when moving from one
+     * step to another.
+     *
+     * @returns {boolean} true if any step's view is still loading.
+     */
+    areSRStepsLoading: function() {
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: This is a better name, and documentation is much better.

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 20]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+        return this.responseView.isRendering ||
+            this.peerView.isRendering ||
+            this.selfView.isRendering ||
+            this.gradeView.isRendering ||
+            this.trainingView.isRendering ||
+            this.staffView.isRendering;
+    },
+
+    /**
+     * Updates text int the Aria live region if all sections are rendered and focuses on the specified ID.
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Typo: "int"

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 21]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+     *
+     * @param {boolean} announceCurrentStatus - true if the current status should be added to the aria-live region
+     *     even if it is now showing
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Do you mean "even if it is NOT showing"?

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 22]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+     * @param {String} stepID - The id of the Step being worked on.
+     * @param {String} usageID  - The Usage id of the Step being worked on.
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: This is the usageID of the xblock, no?

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 23]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+     * @param {boolean} gradeStatus - true if this is a Grade status, false if it is an assessment status.
+     * @param {Object} currentView - Current active view.
+     * @param {String} focusID - The ID of the region to focus on.
+     */
+    announceStatusChangeToSRandFocus: function(announceCurrentStatus,
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Nit: don't like each of these on a separate line.

File: openassessment/xblock/static/js/src/lms/oa_base.js (hunk 21 above)
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Maybe some documentation about the fact that status will be added to the aria-live region regardless of this value, if usageID and focusID are defined and the step is showing.

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 24]
```diff
@@ -68,6 +69,108 @@ OpenAssessment.BaseView.prototype = {
...
+        } else if (announceCurrentStatus) {
+            this.srStatusUpdates.push(text);
+        }
+        if (!this.areSRStepsLoading() && this.srStatusUpdates.length > 0) {
+            this.srReadTexts(this.srStatusUpdates);
+            this.srStatusUpdates = [];
+        }
+    },
+
+    announceErrorMsgToSR: function(message) {
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: This method is oddly named-- it has nothing to do with it being an "error message" per se. Since it is only used one place, I would just inline the code where it is called.

File: openassessment/xblock/static/js/src/lms/oa_grade.js (hunk 16 above)
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_grade.js]: Are you comfortable with announceStatusChangeToSRandFocus checking view.announceStatus directly and setting it to false at the end of the method (since view is already passed into the method)?

**Comment by staubina** [comment]
@cptvitamin Can you take a look at this Sandbox and let me know if the changes are sufficient for TNL-5883 Sub Task https://openedx.atlassian.net/browse/TNL-6241?

**Comment by efischer19** [comment]
All this code looks fine to me. I liked @cahrens's suggestion about a bok choy test addition, but I'll leave the final decision whether to add that up to you.

Consider this 👍'd from me, unless there are changes and you'd like me to re-review.

**Comment by staubina** [comment]
@cahrens I made some minor updates to the bok choy tests.  @cptvitamin is going to take a look today if he has time, most likely tomorrow.

File: test/acceptance/tests.py [hunk 25]
```diff
@@ -1100,7 +1109,6 @@ def test_train_self_peer_staff(self, peer_grades_me):
         # Do staff assessment step
         self.staff_assessment(peer_grades_me)
 
-
```
Comment (efischer19) [inline test/acceptance/tests.py]: This line shouldn't have been deleted - there ought to be 2 blank lines between classes.

**Comment by cahrens** [comment]
👍

**Comment by staubina** [comment]
@efischer19 I made the correction in cf9990f

**Comment by staubina** [comment]
@cptvitamin Please use this Sandbox for testing.  The other one has died.

https://staubina-ora.sandbox.edx.org/

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 26]
```diff
@@ -68,6 +69,95 @@ OpenAssessment.BaseView.prototype = {
...
+     * Add the text messages to the Aria live region (first creating it if necessary).
+     *
+     * @param {string[]} texts
+     */
+    srReadTexts: function(texts) {
+        var readerFeedbackID = '.sr.reader-feedback',
+            $readerFeedbackSelector = $(readerFeedbackID),
+            htmlFeedback = '';
+
+        if ($readerFeedbackSelector.length === 0) {
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Actually, instead of appending dynamically why can't this just always exist (via a template)? I'm concerned that text may not be read initially when it is dynamically injected-- @cptvitamin should know whether or not that is a problem.

But also for simplicity, it seems that it would be better if this region is just always there. I think you could put it in oa_base.html.

**Comment by staubina** [comment]
@cahrens I have made the feedback region a static div in the template.  I am testing acceptance and a11y now.

File: openassessment/xblock/static/js/src/lms/oa_base.js [hunk 27]
```diff
@@ -78,12 +78,6 @@ OpenAssessment.BaseView.prototype = {
             $readerFeedbackSelector = $(readerFeedbackID),
             htmlFeedback = '';
 
-        if ($readerFeedbackSelector.length === 0) {
```
Comment (cahrens) [inline openassessment/xblock/static/js/src/lms/oa_base.js]: Update the doc as well.

**Comment by staubina** [comment]
@cahrens Updated.

**Comment by cahrens** [comment]
👍

**Comment by cptvitamin** [comment]
👍  this rocks.  The feedback was timely and informative.  Loved using it with a screen reader.
